File: packages/gbnf-experiment/data/20260909T142951Z_4ff7cdec/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color } from './colorize.ts';
import type { GraphPointer } from './graph-pointer.ts';

export const getParentStackId = (
  pointer: GraphPointer,
  col: (value: string | number, color: string) => string
): string => {
  const stackIds: string[] = [];
  let parent = pointer.parent;
  while (parent) {
    const { stackId, pathId, stepId } = parent.node.meta;
    stackIds.push(`${stackId},${pathId},${stepId}`);
    parent = parent.parent;
  }
  const arrow = col('<-', Color.GRAY);
  return stackIds.map(stackId => col(stackId, Color.RED)).join(arrow);
};

File: packages/gbnf-experiment/data/20260909T133702Z_8305aa82/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from .colorize import Color


def get_parent_stack_id(pointer, col) -> str:
    stack_ids = []
    parent = pointer.parent
    while parent is not None:
        meta = parent.node.meta
        stack_ids.append(f"{meta['stackId']},{meta['pathId']},{meta['stepId']}")
        parent = parent.parent
    return col("<-", Color.GRAY).join(col(id_, Color.RED) for id_ in stack_ids)
